src/Cache/Cache.php

```diff
@@ -0,0 +1,160 @@
+<?php
+
+namespace BBC\ProgrammesPagesService\Cache;
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use InvalidArgumentException;
+
+class Cache implements CacheInterface
+{
+    /** @var CacheItemPoolInterface */
+    private $cachePool;
+
+    /** @var string */
+    private $prefix;
+
+    /** @var bool */
+    private $flushCacheItems = false;
+
+    private $defaultCacheTimes = [
+        CacheInterface::NONE => -1,
+        CacheInterface::SHORT => 60,
+        CacheInterface::NORMAL => 300,
+        CacheInterface::MEDIUM => 1200,
+        CacheInterface::LONG => 7200,
+        CacheInterface::X_LONG => 86400,
+        CacheInterface::INDEFINITE => 0,
+    ];
+
+    /** @var array */
+    private $cacheTimes;
+
+    public function __construct(
+        CacheItemPoolInterface $cachePool,
+        string $prefix,
+        array $cacheTimes = []
+    ) {
+        $this->cachePool = $cachePool;
+        $this->prefix = $prefix;
+        $this->cacheTimes = array_merge($this->defaultCacheTimes, $cacheTimes);
+    }
+
+    public function getItem(string $key): CacheItemInterface
+    {
+        $key = $this->standardiseKey($key);
+        if ($this->flushCacheItems) {
+            $this->cachePool->deleteItem($key);
+        }
+        return $this->cachePool->getItem($key);
+    }
+
+    /**
+     * @param CacheItemInterface $item
+     * @param int|string $ttl
+     *   TTL in seconds, or a constant from CacheInterface
+     * @return bool
+     */
+    public function setItem(CacheItemInterface $item, $ttl): bool
+    {
+        $ttl = $this->calculateTtl($ttl);
+        $item->expiresAfter($ttl);
+        return $this->cachePool->save($item);
+    }
+
+    /**
+     * IF CALLABLE RETURNS SOMETHING THAT EVALUATES TO EMPTY THE RESULT WILL NOT BE CACHED
+     *
+     * @param string $key
+     * @param int|string $ttl
+     *   TTL in seconds, or a constant from CacheInterface
+     * @param callable $function
+     * @param array|null $arguments
+     * @return mixed
+     */
+    public function getOrSet(string $key, $ttl, callable $function, array $arguments = [])
+    {
+        $cacheItem = $this->getItem($key);
+        if ($cacheItem->isHit() && !$this->flushCacheItems) {
+            return $cacheItem->get();
+        }
+        $ttl = $this->calculateTtl($ttl);
+        $result = call_user_func_array($function, $arguments);
+        if (!empty($result)) {
+            $cacheItem->set($result);
+            $cacheItem->expiresAfter($ttl);
+            $this->cachePool->save($cacheItem);
+        }
+        return $result;
+    }
+
+    /**
+     * Helps you to construct good cache keys by prodding you in the correct direction.
+     * Entirely optional but you are encouraged to use it.
+     *
+     * @param string $className
+     * @param string $functionName
+     * @param \string[] ...$uniqueValues
```

Comment:
All obey the Almighty Intellij.  
![Almighty Intellij](https://upload.wikimedia.org/wikipedia/commons/9/90/Touched_by_His_Noodly_Appendage_HD.jpg)